util/solver.py

```diff
@@ -104,52 +149,89 @@ def train(self, model, train_loader, val_loader, num_epochs=10, log_nth=0):
         - num_epochs: total number of training epochs
         - log_nth: log training accuracy and loss every nth iteration
         """
-        optim = self.optim(model.parameters(), **self.optim_args)
+        self.n_decay_epoch = n_decay_epoch
+        self.decay_factor = decay_factor
+        optim = self.optim(filter(lambda p: p.requires_grad, model.parameters()), **self.optim_args)
         self._reset_histories()
         iter_per_epoch = len(train_loader)
 
         if torch.cuda.is_available():
             model.cuda()
 
-            print('START TRAIN.')
-
-            nIterations = num_epochs*iter_per_epoch
-
-
-            for i in range(num_epochs):
-                for j, (inputs, labels) in enumerate(train_loader, 0):
-
-                    it = i*iter_per_epoch + j
-
-                    inputs = Variable(inputs)
-                    labels = Variable(labels)
-                    (out_pi, out_mu_x, out_mu_y, out_sigma, out_corr) = model(x_train)
-                    loss = self.mdn_loss_function(out_pi, out_mu_x, out_mu_y, out_sigma, out_corr, y_train)
-                    if it%log_nth==0:
-                        print('[Iteration %i/%i] TRAIN loss: %f' % (it,nIterations,loss))
-                        self.train_loss_history.append(loss)
-
-                        optim.zero_grad()
-                        loss.backward()
-                        optim.step()
-
-                        #train_acc = (np.squeeze(np.array(max_index)) == np.squeeze(np.array(labels))).mean()
-                        #self.train_acc_history.append(train_acc)
-
-                        inputs_val = Variable(torch.from_numpy(val_loader.dataset.X))
-                        labels_val = Variable(torch.from_numpy(val_loader.dataset.y))
-                        (out_pi, out_mu_x, out_mu_y, out_sigma, out_corr) = model.forward(inputs_val)
-                        loss_val = self.mdn_loss_function(out_pi, out_mu_x, out_mu_y, out_sigma, out_corr, y_train)
-
-                        #val_acc = (np.squeeze(np.array(max_index_val)) == np.squeeze(np.array(labels_val))).mean()
-                        #self.val_acc_history.append(val_acc)
-                        #print('[Epoch %i/%i] TRAIN acc/loss: %f/%f' % (i,num_epochs,train_acc, loss))
-                        #print('[Epoch %i/%i] VAL acc/loss: %f/%f' % (i,num_epochs,val_acc, loss_val))
-                        print('[Epoch %i/%i] TRAIN loss: /%f' % (i,num_epochs,loss))
-                        print('[Epoch %i/%i] VAL loss: %f' % (i,num_epochs,loss_val))
-
-
-            print('FINISH.')
+        print('START TRAIN.')
+
+        nIterations = num_epochs*iter_per_epoch
+
+
+        for i in range(num_epochs):
```

Comment:
for i in range(1,num_epochs+1):

This is a conflict with master. You need to start from 1 since zero makes problem in the lr_decay function. It is also possible to change the function instead of this.